### ISSUE TITLE
Seeking Help/Question: Best practice for fine-tuning LLM with Lora but with addtional parameters

### ISSUE THREAD

**Issue by zhouhanxie** [huggingface/peft#1632]
Hi PEFT Library, thanks for the cool project!

So I recently run into a use case where I want to fine-tune a LLM with Lora, but also some additional parameters. I'm currently using the Huggingface eco system for training. I wonder what would be the best practice for saving the additional parameters I want to have?

Specifically, I'm training a model with a PLM component. Only the PLM component is lora-adapted, but there is some extra modules that are being fine-tuned along with the lora adapters. The desired behavior is I want to save all parts that require gradients, not just the lora adapter, when calling model.save().

Here is a simplified example of the model I'm working with:

```
from transformers import PreTrainedModel, LlamaModel
from transformers.modeling_utils import PretrainedConfig
from transformers.modeling_outputs import SequenceClassifierOutput
import torch

class MyModelConfig(PretrainedConfig)
       # .... some useful custom config settings...

class MyABitComplexExtraModel(torch.nn.Module):
       # a somewhat complex pytorch model that takes llama's output hidden state
       # and map it to label

class MyModel(PreTrainedModel):
        def __init__(config:MyModelConfig):
                 # ...... some intializations
                 self.llm = LlamaModel.from_pretrained(config.llama_path) # <-- the llm I want to do LoRa on
                 self.extra_module = MyABitComplexExtraModel()
        def forward(self, llamamodel_inputs, labels):
                 llama_output = self.llm(**llamamodel_inputs)
                 logits = self.extra_module(llama_output)
                 if labels:
                           loss = ....
                           return SequenceClassifierOutput(loss=loss, logits=logits)
                 return logits

# preparing model for training
model = MyModel(config)

from peft import get_peft_model
for param in model.llm.parameters():
     param.requires_grad = False
model.llm = get_peft_model(model.llm, peft_config) #<-- note that the model.extra_module is still being tuned here
```

Now, when I later train this model using a standard HuggingFace trainer, is there any simply way to make sure every time I call model.save(), not only the adapter but as well as the "extra_module" is saved? 

I looked around and currently the only way seems to be somehow inherit the "BaseTunerLayer" in "MyABitComplexExtraModel"? However, this probably require making the whole "MyABitComplexExtraModel" PEFT-compatible, which probably requires implementing its own configs, etc... 

I wonder if there'd be easier way to do this, given this is experimental code and will be heavily modified... Thanks!

**Comment by BenjaminBossan** [comment]
> Now, when I later train this model using a standard HuggingFace trainer, is there any simply way to make sure every time I call model.save(), not only the adapter but as well as the "extra_module" is saved?

I assume you mean `save_pretrained`? No, unfortunately, I don't think there is an easy way to do this.

You could write your own custom functions to save and load those extra parameters, but probably that's not quite enough for you because you want to use `Trainer`, which isn't aware of those functions (e.g. when creating checkpoints).

What you could try is to create very light weight wrapper, something along the lines of:

```python
class MyCustomSaveWrapper(nn.Module):
    # wraps a peft model and overrides save_pretrained and from_pretrained,
    # otherwise, it is a pass-through
    def __init__(self, model):
        super().__init__()
        self.model = model

    def __getattr__(self, name):
        return getattr(self.model, name)

    def save_pretrained(self, *args, **kwargs):
        # ensure that extra params are saved
        # take care of removing the extra 'model.' prefix added by this wrapper
        ...
```

There may be some more steps involved than what I outlined here, but it could be worth a try.

**Comment by zhouhanxie** [comment]
Ok, thanks! Yeah, that do sounds like a great idea... (Though still hope there could be some sort of mixins we can easily inherit...)

Closing and thanks for the quick response!
